libraries/AP_EFI/AP_EFI_Serial_FH.cpp

```diff
@@ -0,0 +1,155 @@
+          else if (sizeof(pkt) == ind) return (0xD == data.eop && data.sum == sum8(pkt, 48));
+        }
+      }
+    }
+    return false;
+  }
+
+static unsigned char boza = 0;
+#include "../AP_ICEngine/AP_ICEngine.h"
+
+void AP_EFI_Serial_FH::update()
+{
+    const uint32_t now = AP_HAL::millis();
+    boza++;
+    if (process(now)) {
+        boza=0x66;
+        internal_state.spark_dwell_time_ms = 123.45f;        
+        // RPM and Engine State
+        internal_state.engine_speed_rpm = data.rpm;
+        if (internal_state.engine_speed_rpm > 0) {
+          internal_state.engine_state = Engine_State::RUNNING;
+        } else {
+          internal_state.engine_state = Engine_State::STOPPED;
+        }
+        // EFI ECU power [volts]
+        internal_state.ignition_voltage = tenth(data.ivolt);
+        internal_state.atmospheric_pressure_kpa = tenth(data.ivolt);
+        //0=-273.15 273.15=0
+        internal_state.intake_manifold_temperature = degC_to_Kelvin(data.tempOut);
+        internal_state.intake_manifold_pressure_kpa = tenth(data.FP);
+        
+        auto &c = internal_state.cylinder_status;
+        // CHT temp 1/2
+        c.cylinder_head_temperature  = degC_to_Kelvin(MAX(data.temp[0], data.temp[1]));
+        c.cylinder_head_temperature2 = degC_to_Kelvin(MIN(data.temp[0], data.temp[1]));
+        // EGT temp 3/4
+        c.exhaust_gas_temperature = degC_to_Kelvin(MAX(data.temp[2], data.temp[3]));
+        c.exhaust_gas_temperature2 = degC_to_Kelvin(MIN(data.temp[2], data.temp[3]));
+        // PMU/GEN temp
+        internal_state.coolant_temperature = degC_to_Kelvin(data.temp[4]);
+        internal_state.oil_temperature = degC_to_Kelvin(data.temp[5]);
+        // Injection timing
+        c.injection_time_ms = thousnd(data.ilus);            
+        // Lambda
+        c.lambda_coefficient = tenth(data.LV);
+        // Throttle Position [%]
+        internal_state.throttle_position_percent = data.tps;
+
+        internal_state.fuel_pressure = tenth(data.FP); // Fuel_Pressure_Status::OK;
+        // Fuel consumption
+        internal_state.fuel_consumption_rate_cm3pm = convertFuelConsumption((float)data.CFCHPL/1000.f);
+        // Integrate fuel consumption
+        if (internal_state.engine_speed_rpm > RPM_THRESHOLD) {
+            //const float duty_cycle = (internal_state.cylinder_status.injection_time_ms * internal_state.engine_speed_rpm) / 600.0f;
+            //internal_state.fuel_consumption_rate_cm3pm = duty_cycle * get_coef1() - get_coef2();
+            internal_state.estimated_consumed_fuel_volume_cm3 += internal_state.fuel_consumption_rate_cm3pm * (now - internal_state.last_updated_ms) / 60000.0f;
+        } else {
+            //internal_state.fuel_consumption_rate_cm3pm = 0;
+        }
+        internal_state.last_updated_ms = now;
+        copy_to_frontend();
+    }
+    if (now - last_request_ms > 200) {
+        last_request_ms = now;
+        port->discard_input();

```

Comment:
You're not killing your buffer pointer here, which you should.

I suggest killing the current parser.  It's not a complicated protocol.

Read an entire packet at a time, or none at all.  If you ever see a number of bytes available to read you don't recognise, discard the input.  Slurp the bytes into your buffer, validate the packet, call process.